### PR TITLE
Add OpenAPI json rendering

### LINKED ISSUE #1499: Add .toJson methods to OpenAPI data types
**Is your feature request related to a problem? Please describe.**

The `zio.http.api.openapi.*` family of data types is an ADT to describe OpenAPI documentation. The ADT is strongly typed and incorporates data types from `zio.http.model`, as well as ZIO and the Scala standard library.

While useful, this ADT cannot currently be translated into rendered JSON, which impedes other tickets.

**Describe the solution you'd like**

Add `def toJson: String` methods to all OpenAPI data types, which directly render the OpenAPI data types to JSON, using some utility JSON class, but without introducing a dependency on a third-party JSON library (which is unnecessary for pure rendering).

**Describe alternatives you've considered**

Using `zio-json` to perform the serialization, but that forces us to use a schema that may not be ideal, and makes the code fragile to changes, as well as introducing a direct dependency on a JSON library for both client and server use cases.

**Additional context**

The [OpenAPI specification](https://swagger.io/specification/) has details on how the OpenAPI information should be encoded in JSON or YAML formats.

### PR DESCRIPTION
**Pull request by 987Nabil**
some minor changes to OpenAPI data types,
to match the specs (one could say bugfixes)

fixes #1499

### CODE REVIEW

File: zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    case l: Iterable[Any]                  => renderIterable(l)
+    case (k, v)                            => s"{${renderKey(k)}:${renderValue(v)}}"
+    case u: URI                            => s""""${u.toString}""""
+    case p: Product if p.productArity == 0 => renderSingleton(p)
+    case other                             => s""""${other.toString}""""
+  }
+
+  private def renderDoc(d: Doc) =
+    // todo render markdown
+    s""""${d.toPlaintext(color = false)}"""".replace("\n", "<br/>")
```
Comment (jdegoes) [inline zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala]: In some cases, OpenAPI lets us embed CommonMark (or html), while other places, we must use plaintext. However, `renderValue` always uses plaintext?

File: zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      throw new IllegalArgumentException("Duplicate field names")
+    } else {
+      val fields = fieldsIt
+        .filterNot(_._2 == None) // We don't render empty fields
+        .map { case (name, value) => s""""$name":${renderValue(value)}""" }
+      s"{${fields.mkString(",")}}"
+    }
+  }
+
+  private def renderValue(value: Any): String = value match {
```
Comment (jdegoes) [inline zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala]: Can we make this more type safe, i.e. define `trait Renderable[A]` or something?

File: zio-http/src/main/scala/zio/http/api/openapi/OpenAPI.scala [hunk 3]
```diff
@@ -377,26 +483,52 @@ object OpenAPI {
...
     sealed trait QueryStyle
 
-    case object Matrix         extends PathStyle
-    case object Label          extends PathStyle
-    case object Simple         extends PathStyle
-    case object Form           extends QueryStyle
-    case object SpaceDelimited extends QueryStyle
-    case object PipeDelimited  extends QueryStyle
-    case object DeepObject     extends QueryStyle
+    object QueryStyle {
```
Comment (jdegoes) [inline zio-http/src/main/scala/zio/http/api/openapi/OpenAPI.scala:518]: Love the cleanup!

**Review by jdegoes — CHANGES_REQUESTED** [review]
Looks fantastic! Just a few comments.

File: zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala (hunk 1 above)
Comment (987Nabil) [reply zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala]: The plaintext rendering renders new lines. Which is not valid inside a json string, so I have to escape it.
The html rendering of doc, renders a full html doc, not just a snippet. Including a big block of styling. So it is not usable.

File: zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala (hunk 2 above)
Comment (987Nabil) [reply zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala]: I'll do that.

File: zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala (hunk 1 above)
Comment (987Nabil) [reply zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala]: Example HTML output for one link.

```
+{"doc":"<html><head><style>
      +        h1 {
      +            color: rgb(36, 41, 46);
      +            font-weight: 600;
      +            line-height: 1.25;
      +            margin-bottom: 16px;
      +            font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji, Segoe UI Symbol;
      +        }
      +
      +        h2 {
      +            font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji, Segoe UI Symbol;
      +            font-size: 24px;
      +            letter-spacing: 0px;
      +            word-spacing: 2px;
      +            color: rgb(36, 41, 46);
      +            font-weight: 600;
      +        }
      +
      +        h3 {
      +            font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji, Segoe UI Symbol;
      +            font-size: 21px;
      +            letter-spacing: 0px;
      +            word-spacing: 2px;
      +            color: rgb(36, 41, 46);
      +            font-weight: 700;
      +        }
      +
      +        p {
      +            font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji, Segoe UI Symbol;
      +            color: #24292e;
      +        }
      +
      +        .error {
      +            font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji, Segoe UI Symbol;
      +            color: #24292e;
      +        }
      +
      +        a {
      +            border: 0;
      +            color: rgb(189, 39, 26);
      +            text-decoration: none;
      +            font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji, Segoe UI Symbol;
      +            font-size: inherit;
      +            font-size: 100%;
      +            margin: 0;
      +            padding: 0;
      +            vertical-align: baseline;
      +        }
      +
      +        a:hover {
      +            color: rgb(0, 0, 0);
      +        }
      +
      +        pre {
      +          background-color: rgba(27, 31, 35, .05);
      +        }
      +
      +        code {
      +            border-radius: 3px;
      +            color: rgb(36, 41, 46);
      +            font-family: SFMono-Regular, Menlo, Monaco, Consolas, Liberation Mono, Courier New, monospace;
      +            font-size: 85%;
      +            margin: 0;
      +        }
      +    </style></head><body><p><a href="https://google.com">https://google.com</a></p></body></html>"}
```

Comment (987Nabil) [reply zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala]: And this is also not a supported [CommanMark ](https://spec.commonmark.org/0.27/#html-blocks) html subset.

File: zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala (hunk 2 above)
Comment (987Nabil) [reply zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala]: @jdegoes I just realized, I am not sure I can. Please take a look at the `Link` class. It has fields of type any.

File: zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala (hunk 1 above)
Comment (jdegoes) [reply zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala]: Would you mind deleting all that unnecessary `<html>` and make it so it renders as inline HTML?

File: zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala (hunk 2 above)
Comment (jdegoes) [reply zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala]: I think we need to change that, don't we? What is the OpenAPI spec say can be inside `Link`?

Comment (987Nabil) [reply zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala]: The spec says actually `Any`. But maybe we have to be more restrictive? Also, what/how would you read this, if you want to impl. #1520

Comment (987Nabil) [reply zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala]: https://swagger.io/specification/#Link_Object

**Comment by 987Nabil** [comment]
@jdegoes I just ignored the fields of `Link` with type `Any` for the json rendering. Maybe that is ok for now? And we can defer what to do with these fields, once we impl. #1520 and #1498 ?

File: zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala (hunk 2 above)
Comment (jdegoes) [reply zio-http/src/main/scala/zio/http/api/openapi/JsonRenderer.scala]: So we are missing a `LinkObject` which is a `Map[String, LinkObject | Reference]`

**Review by jdegoes — APPROVED** [review]

**Comment by jdegoes** [comment]
@987Nabil Follow-up PRs to address the rendering and link object changes would be appreciated!

**Comment by jdegoes** [comment]
Thanks for your patience and hard work getting this one in.

**Pull request merged**